### PR TITLE
Consistent get_<DATA_ASSET> permissions - S3_Datasets

### PR DESCRIPTION
**Pull request by dlpzx**
### Feature or Bugfix
- Feature/Bugfix

### Detail

This PR is the first part of a series and only handles S3_Datasets and its child components Tables and Folders

#### Current implementation
Most API calls on a particular resource are restricted by GET_RESOURCE permissions. But for resources that are indexed in the Catalog, some metadata is considered public as it is useful for data consumers to discover and understand the data assets.  Users will click on these resources from the Catalog view and call one of the following API calls:
- getDataset
- getDatasetStorageLocation
- getDatasetTable
- getDashboard
- getRedshiftDataset
- getRedshiftTable

From the above list, initially all are decorated with resource_permission checks except for getDataset and getDatasetTable. 

#### Target implementation
- Public information should be available for data consumers to explore, that means that we first need to remove the resource_permission checks from the list of APIs.
- Not all information is public, to get AWS information and other restricted metadata we still need to verify GET_X permissions on the resource. 
- For restricted metadata, we should provide default messages that do not break the frontend

In addition in this PR some unused fields are removed and `syncTables` is simplified to return an integer with the count of synced tables

### Testing
For each of the following I tested with a user with GET permissions and without GET permissions. FE views render and show information or unauthorized to view info placeholder
- [X] Dataset View, Dataset Edit Form and list Datasets
- [x] Dataset Data tab with list of tables and folders
- [X] Table view, Table Edit
- [X] Folder view and Folder Edit

Other checks
- [x] Complete share request
- [X] With requester check table and folder and view the details of the account...
- [X] Worksheets query with owned table
- [X] Worksheets query with shared table
- [X] Crawl dataset - correct S3 path
- [X] Sync tables - correct params

### Security
Please answer the questions below briefly where applicable, or write `N/A`. Based on
[OWASP 10](https://owasp.org/Top10/en/).

- Does this PR introduce or modify any input fields or queries - this includes
fetching data from storage outside the application (e.g. a database, an S3 bucket)?
  - Is the input sanitized?
  - What precautions are you taking before deserializing the data you consume?
  - Is injection prevented by parametrizing queries?
  - Have you ensured no `eval` or similar functions are used?
- Does this PR introduce any functionality or component that requires authorization?
  - How have you ensured it respects the existing AuthN/AuthZ mechanisms?
  - Are you logging failed auth attempts?
- Are you using or adding any cryptographic features?
  - Do you use a standard proven implementations?
  - Are the used keys controlled by the customer? Where are they stored?
- Are you introducing any new policies/roles/users?
  - Have you used the least-privilege principle? How?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: backend/dataall/modules/s3_datasets/services/dataset_location_service.py
Comment (petrkalos) [inline backend/dataall/modules/s3_datasets/services/dataset_location_service.py:163]: I'd let this exception bubble up and appear in the error section of the gql response

File: backend/dataall/modules/s3_datasets/services/dataset_location_service.py
Comment (petrkalos) [inline backend/dataall/modules/s3_datasets/services/dataset_location_service.py:151]: What happens if we return the entire object? Would GQL fail because of the extra fields or will only return those that have mappings?
If that's not possible then maybe something like 
`return {field.name: getattr(dataset, field.name, None) for field in DatasetRestrictedInformation.fields}`
 would work better as we won't have to maintain those mappings manually. You can also export it to global method and reuse it an all places

**Review by petrkalos — CHANGES_REQUESTED** [review]

File: backend/dataall/modules/s3_datasets/services/dataset_location_service.py
Comment (dlpzx) [reply backend/dataall/modules/s3_datasets/services/dataset_location_service.py:163]: Why? The end user is authorized to get the object, but some fields are restricted

File: backend/dataall/modules/s3_datasets/services/dataset_location_service.py
Comment (dlpzx) [reply backend/dataall/modules/s3_datasets/services/dataset_location_service.py:151]: You are right. Let me change it a bit

File: backend/dataall/modules/s3_datasets/services/dataset_location_service.py
Comment (petrkalos) [reply backend/dataall/modules/s3_datasets/services/dataset_location_service.py]: They will get the top level object but an exception for the "restricted" subsection, that way it is more consistent to what the rest of the sub-resolvers are doing

Comment (dlpzx) [reply backend/dataall/modules/s3_datasets/services/dataset_location_service.py]: I'll work on it, it is true that for consistency it would be better to not retrieve anything

Comment (dlpzx) [reply backend/dataall/modules/s3_datasets/services/dataset_location_service.py]: I have submitted a new version and tested the following with a user with GET permissions and without GET permissions:
- Dataset view and all tabs
- Table view and tabs
- Folder view
- Worksheets dropdowns

File: backend/dataall/modules/s3_datasets/services/dataset_service.py
Comment (petrkalos) [inline backend/dataall/modules/s3_datasets/services/dataset_service.py:4]: unused

I usually do a `Code -> Optimize Imports` before commit to automatically remove unused imports

File: backend/dataall/modules/s3_datasets/services/dataset_location_service.py
Comment (petrkalos) [inline backend/dataall/modules/s3_datasets/services/dataset_location_service.py:1]: unused

File: backend/dataall/modules/s3_datasets/services/dataset_location_service.py
Comment (petrkalos) [inline backend/dataall/modules/s3_datasets/services/dataset_location_service.py:7]: unused

File: backend/dataall/modules/s3_datasets/services/dataset_table_service.py
Comment (petrkalos) [inline backend/dataall/modules/s3_datasets/services/dataset_table_service.py:2]: unused

File: frontend/src/modules/Folders/components/FolderOverview.js
Comment (petrkalos) [inline frontend/src/modules/Folders/components/FolderOverview.js:31]: * This is already handled in [ObjectMetadata](https://github.com/data-dot-all/dataall/blob/eca354e8df8189c2fcf31732579def69bad90bfb/frontend/src/design/components/ObjectMetadata.js#L127).
* I think it's better not to print region at all than printing "Region -> UNAUTHORIZED"

ditto for the rest of your special handling

**Review by petrkalos — CHANGES_REQUESTED** [review]

File: frontend/src/modules/Folders/components/FolderOverview.js
Comment (dlpzx) [reply frontend/src/modules/Folders/components/FolderOverview.js]: In the latest commit 4b675f3 most views have been simplified except for the tables and folder view where table names are presented in the rows.

**Review by petrkalos — APPROVED** [review]

**Review by petrkalos — APPROVED** [review]

File: backend/dataall/modules/s3_datasets/api/dataset/types.py
Comment (noah-paige) [inline backend/dataall/modules/s3_datasets/api/dataset/types.py:33]: nit: A call out that these 2 fields (`S3BucketName` and `GlueDatabaseName`) are a part of index in opensearch (`backend/dataall/modules/s3_datasets/indexers/dataset_indexer.py`)  - not sure really how "restricted" they are treated if that information can be found elsewhere

File: backend/dataall/modules/s3_datasets/api/dataset/types.py
Comment (noah-paige) [inline backend/dataall/modules/s3_datasets/api/dataset/types.py:49]: nit: making note we can remove the following from data model if we wanted to clean up sometime in the future

- bucketCreated
- glueDatabaseCreated
- iamAdminRoleCreated
- lakeformationLocationCreated
- bucketPolicyCreated

File: backend/dataall/modules/s3_datasets/api/table/types.py
Comment (noah-paige) [inline backend/dataall/modules/s3_datasets/api/table/types.py:30]: similar to call out above that `GlueDatabaseName` is a part of index in opensearch (`backend/dataall/modules/s3_datasets/indexers/table_indexer.py`) - not sure really how "restricted" they are treated if that information can be found elsewhere

++ to note I would think best if anything would be to restrict on open search and keep the other server side restricted logic the same

File: frontend/src/modules/Folders/views/FolderView.js
Comment (noah-paige) [inline frontend/src/modules/Folders/views/FolderView.js:236]: nit: update the error dispatch to dispatch all errors (1 or more)

File: frontend/src/modules/S3_Datasets/components/DatasetFolders.js
Comment (noah-paige) [inline frontend/src/modules/S3_Datasets/components/DatasetFolders.js:73]: nit: dispatch all errors

File: frontend/src/modules/S3_Datasets/components/DatasetTables.js
Comment (noah-paige) [inline frontend/src/modules/S3_Datasets/components/DatasetTables.js:83]: nit: dispatch all errors

File: frontend/src/modules/S3_Datasets/components/DatasetUpload.js
Comment (noah-paige) [inline frontend/src/modules/S3_Datasets/components/DatasetUpload.js:156]: nit: i know tab is limited to admin but any need to also protect this with `emptyPrintUnauthorized` or similar?

File: frontend/src/modules/S3_Datasets/components/DatasetStartCrawlerModal.js
Comment (noah-paige) [inline frontend/src/modules/S3_Datasets/components/DatasetStartCrawlerModal.js:135]: nit: i know tab is limited to admin but any need to also protect this with `emptyPrintUnauthorized` or similar?

File: backend/dataall/modules/s3_datasets/api/dataset/types.py
Comment (noah-paige) [inline backend/dataall/modules/s3_datasets/api/dataset/types.py:39]: did we also want to include the following share expiration / config related information as part of restricted?
```
        gql.Field(name='autoApprovalEnabled', type=gql.Boolean),
        gql.Field(name='enableExpiration', type=gql.Boolean),
        gql.Field(name='expirySetting', type=gql.String),
        gql.Field(name='expiryMinDuration', type=gql.Integer),
        gql.Field(name='expiryMaxDuration', type=gql.Integer),
```

File: frontend/src/modules/S3_Datasets/views/DatasetView.js
Comment (noah-paige) [inline frontend/src/modules/S3_Datasets/views/DatasetView.js:164]: nit: expose all errors

File: frontend/src/modules/Tables/views/TableView.js
Comment (noah-paige) [inline frontend/src/modules/Tables/views/TableView.js:225]: nit: expose all errors

File: frontend/src/utils/helpers/emptyPrintUnauthorized.js
Comment (noah-paige) [inline frontend/src/utils/helpers/emptyPrintUnauthorized.js:2]: very nit: would it be a better UX to have the Unauth string be more like `******` to represent redacted/restricted info rather than a replacement string of `UNAUTHORIZED_INFO`

File: tests/modules/s3_datasets_shares/conftest.py
Comment (noah-paige) [inline tests/modules/s3_datasets_shares/conftest.py:262]: nit: any reason handled this way here and updated `def table` factory for a similar/same required change at `tests/modules/s3_datasets/conftest.py`

**Comment by noah-paige** [comment]
All minor enhancmenets / nits - the only pending comment I have before approval is regarding restrictions to additional dataset config info (re: https://github.com/data-dot-all/dataall/pull/1727#discussion_r1871981295) cc: @dlpzx

File: backend/dataall/modules/s3_datasets/api/dataset/types.py
Comment (dlpzx) [reply backend/dataall/modules/s3_datasets/api/dataset/types.py:49]: I did not want to add a migration script because backfilling it in 2.6.2 was going to be a pain; but totally agree on the cleanup

File: backend/dataall/modules/s3_datasets/api/dataset/types.py
Comment (dlpzx) [reply backend/dataall/modules/s3_datasets/api/dataset/types.py:39]: Initially yes, but this information is fetched when opening a share request. At the end the user needs it for the share request creation

File: backend/dataall/modules/s3_datasets/api/table/types.py
Comment (dlpzx) [reply backend/dataall/modules/s3_datasets/api/table/types.py:30]: I think this might be the solution, restricting it also in OpenSearch

File: frontend/src/modules/S3_Datasets/components/DatasetUpload.js
Comment (dlpzx) [reply frontend/src/modules/S3_Datasets/components/DatasetUpload.js:156]: This PR has gone back and forth. At the beginning I added checks everywhere, but finally I opted to use it only when it will be used. The same as if in the backend an if condition is not needed then we don't add it

File: frontend/src/utils/helpers/emptyPrintUnauthorized.js
Comment (dlpzx) [reply frontend/src/utils/helpers/emptyPrintUnauthorized.js:2]: Done!

File: frontend/src/modules/S3_Datasets/views/DatasetView.js
Comment (dlpzx) [reply frontend/src/modules/S3_Datasets/views/DatasetView.js:164]: Same as with Dashboards. I don't know if adding more error banners is going to give users more useful info. If the top level query fails, then the errors that we might find in the nested resolvers are very unpredictable. For example if get_Dataset fails because the dataset is not found, the restricted is going to fail with something as 'URI is null'; which might be misleading.

File: tests/modules/s3_datasets_shares/conftest.py
Comment (dlpzx) [reply tests/modules/s3_datasets_shares/conftest.py:262]: This is not very elegant, agreed. The issue is that the `dataset_fixture` parameter takes in this case the value of the gql Dataset type. While the table factory accepts db S3Dataset model

File: frontend/src/modules/S3_Datasets/views/DatasetView.js
Comment (noah-paige) [reply frontend/src/modules/S3_Datasets/views/DatasetView.js:164]: is it guaranteed the first error in the dispatch always the top level?

**Comment by noah-paige** [comment]
@dlpzx - I think will have to merge latest from OS, resolve conflicts, and make changes to `test_permissions` for the changes in auth checks / nested resolvers 

Then good to go

File: tests/modules/s3_datasets_shares/conftest.py
Comment (noah-paige) [reply tests/modules/s3_datasets_shares/conftest.py:262]: I am not sure I follow - is it not the exact same dataset and table factories used in `tests/modules/s3_datasets/conftest.py` but in that conftest we define Table factor as

```
@pytest.fixture(scope='module', autouse=True)
def table(db):
    cache = {}

    def factory(dataset: S3Dataset, name, username) -> DatasetTable:
        key = f'{dataset.datasetUri}-{name}'
        if cache.get(key):
            return cache.get(key)
        with db.scoped_session() as session:
            table = DatasetTable(
                name=name,
                label=name,
                owner=username,
                datasetUri=dataset.datasetUri,
                GlueDatabaseName=dataset.restricted.GlueDatabaseName,
                GlueTableName=name,
                region=dataset.restricted.region,
                AWSAccountId=dataset.restricted.AwsAccountId,
                S3BucketName=dataset.restricted.S3BucketName,
                S3Prefix=f'{name}',
            )
...
```

(note the way we set GlueDatabaseName, region, S3BucketName, and S3BucketName)

**Review by noah-paige — APPROVED** [review]

**Pull request merged**
